lib/src/rules/use_function_declaration_instead_variable.dart

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+library linter.src.rules.use_function_declaration_instead_variable;
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'Use a function declaration to bind a function to a name.';
+
+const _details = r'''
+
+**DO** use a function declaration to bind a function to a name.
+
+**BAD:**
+```
+void main() {
+  var localFunction = () {
+    ...
+  };
+}
+```
+
+**GOOD:**
+```
+void main() {
+  localFunction() {
+    ...
+  }
+}
+```
+
+''';
+
+class UseFunctionDeclarationInsteadVariable extends LintRule {
+  _Visitor _visitor;
+  UseFunctionDeclarationInsteadVariable()
+      : super(
+            name: 'use_function_declaration_instead_variable',
+            description: _desc,
+            details: _details,
+            group: Group.style) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitAssignmentExpression(AssignmentExpression node) {
+    if (node.rightHandSide.bestType is FunctionType) {
+      rule.reportLint(node);
+    }
+  }
+
+  @override
+  visitVariableDeclaration(VariableDeclaration node) {
+    if (node.initializer != null && node.initializer.bestType is FunctionType) {
```

Comment:
I don't think you want to test the type of the initializer here at all. I think you want to ask `node.initializer is FunctionExpression`.

Otherwise you'll end up creating a lint for something like
```
void doIt(Function f) {
  var x = f; // OK
  x(this);
}
```